File: 1. Creational Design Patterns/Abstract Factory/Java/AbstractFactoryTemplate.java
// Abstract Factory Pattern – Java Template
// Fill concrete products and factories for your scenario.

interface ProductA {}
interface ProductB {}

class ConcreteProductA1 implements ProductA {}
class ConcreteProductB1 implements ProductB {}

interface AbstractFactory {
    ProductA createProductA();
    ProductB createProductB();
}

class ConcreteFactory1 implements AbstractFactory {
    public ProductA createProductA() { return new ConcreteProductA1(); }
    public ProductB createProductB() { return new ConcreteProductB1(); }
}

public class AbstractFactoryTemplate {
    public static void main(String[] args) {
        // Client code here
        AbstractFactory factory = new ConcreteFactory1();
    }
}

File: 1. Creational Design Patterns/Abstract Factory/Kotlin/AbstractFactoryTemplate.kt
// Abstract Factory Pattern – Kotlin Template

interface ProductA
interface ProductB

class ConcreteProductA1 : ProductA
class ConcreteProductB1 : ProductB

interface AbstractFactory {
    fun createProductA(): ProductA
    fun createProductB(): ProductB
}

class ConcreteFactory1 : AbstractFactory {
    override fun createProductA() = ConcreteProductA1()
    override fun createProductB() = ConcreteProductB1()
}

fun main() {
    val factory: AbstractFactory = ConcreteFactory1()
    // val productA = factory.createProductA()
}
